src/Nethermind/Nethermind.Facade/Eth/RpcTransaction/TransactionForRpc.cs

```diff
@@ -0,0 +1,154 @@
+// SPDX-FileCopyrightText: 2024 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using Nethermind.Core;
+using Nethermind.Core.Crypto;
+using Nethermind.Int256;
+
+namespace Nethermind.Facade.Eth.RpcTransaction;
+
+/// <summary>
+/// Base class for all output Nethermind RPC Transactions.
+/// </summary>
+/// <remarks>
+/// Input:
+/// <para>JSON -> <see cref="TransactionForRpc"></see> (through <see cref="JsonConverter"/>, a registry of [<see cref="TxType"/> => <see cref="TransactionForRpc"/> subtypes)</para>
+/// <para><see cref="TransactionForRpc"/> -> <see cref="Transaction"/> (through an overloaded <see cref="ToTransaction"> method)</para>
+/// Output:
+/// <para><see cref="Transaction"/> -> <see cref="TransactionForRpc"/> (through <see cref="TransactionConverter"/>, a registry of [<see cref="TxType"/> => <see cref="IFromTransaction{T}"/>)</para>
+/// <para><see cref="TransactionForRpc"/> -> JSON (Derived by <c>System.Text.JSON</c> using the runtime type)</para>
+/// </remarks>
+[JsonConverter(typeof(JsonConverter))]
+public abstract class TransactionForRpc
+{
+    [JsonIgnore(Condition = JsonIgnoreCondition.Never)]
+    public virtual TxType? Type => null;
+
+    [JsonIgnore(Condition = JsonIgnoreCondition.Never)]
+    public Hash256? Hash { get; set; }
+
+    [JsonIgnore(Condition = JsonIgnoreCondition.Never)]
+    public long? TransactionIndex { get; set; }
+
+    [JsonIgnore(Condition = JsonIgnoreCondition.Never)]
+    public Hash256? BlockHash { get; set; }
+
+    [JsonIgnore(Condition = JsonIgnoreCondition.Never)]
+    public long? BlockNumber { get; set; }
+
+    [JsonConstructor]
+    public TransactionForRpc() { }
+
+    public TransactionForRpc(Transaction transaction, int? txIndex = null, Hash256? blockHash = null, long? blockNumber = null)
+    {
+        Hash = transaction.Hash;
+        TransactionIndex = txIndex;
+        BlockHash = blockHash;
+        BlockNumber = blockNumber;
+    }
+
+    public virtual Transaction ToTransaction()
+    {
+        return new Transaction
+        {
+            Type = Type ?? default,
+        };
+    }
+
+    public abstract void EnsureDefaults(long? gasCap);
+
+    public class JsonConverter : JsonConverter<TransactionForRpc>
+    {
+        private readonly Type[] _transactionTypes = new Type[Transaction.MaxTxType + 1];
+
+        // TODO: Refactoring transition code
+        public JsonConverter()
+        {
+            RegisterTransactionType(TxType.Legacy, typeof(LegacyTransactionForRpc));
+            RegisterTransactionType(TxType.AccessList, typeof(AccessListTransactionForRpc));
+            RegisterTransactionType(TxType.EIP1559, typeof(EIP1559TransactionForRpc));
+            RegisterTransactionType(TxType.Blob, typeof(BlobTransactionForRpc));
+            // TODO: Add Optimism:
+            // RegisterTransactionType(TxType.DepositTx, typeof(RpcOptimismTransaction))
+        }
```

Comment:
For ease of usage `TransactionForRpc` uses this converter which initializes itself with known transaction types. I think our options are

- Manually register an instance of this converter with manually registered known subclasses. This makes testing harder since we need to add this converter instance to each test that uses the default `EthereumJsonSerializer`.
- Assembly scan all subclasses of `TransactionForRpc` and register them in the converter default constructor, keeping the `[JsonConverter(typeof(JsonConverter))]` annotation. This approach is the one used on RLP (de)serialization.

Reply:
I changed it to register in one place